arangod/Pregel/OutgoingCache.cpp

```diff
@@ -100,50 +102,51 @@ void ArrayOutCache<M>::flushMessages() {
   auto& server = this->_config->vocbase()->server();
   auto const& nf = server.template getFeature<arangodb::NetworkFeature>();
   network::ConnectionPool* pool = nf.pool();
-
   network::RequestOptions reqOpts;
   reqOpts.database = this->_config->database();
   reqOpts.skipScheduler = true;
 
   std::vector<futures::Future<network::Response>> responses;
-  for (auto const& it : _shardMap) {
-    PregelShard shard = it.first;
-    std::unordered_map<std::string, std::vector<M>> const& vertexMessageMap =
-        it.second;
+  for (auto const& [shard, vertexMessageMap] : _shardMap) {
     if (vertexMessageMap.size() == 0) {
       continue;
     }
 
-    VPackBuffer<uint8_t> buffer;
-    VPackBuilder data(buffer, &options);
-    data.openObject();
-    data.add(Utils::senderKey, VPackValue(ServerState::instance()->getId()));
-    data.add(Utils::executionNumberKey,
-             VPackValue(this->_config->executionNumber().value));
-    data.add(Utils::globalSuperstepKey, VPackValue(gss));
-    data.add(Utils::shardIdKey, VPackValue(shard));
-    data.add(Utils::messagesKey, VPackValue(VPackValueType::Array, true));
-    for (auto const& vertexMessagePair : vertexMessageMap) {
-      data.add(VPackValue(vertexMessagePair.first));      // key
-      data.add(VPackValue(VPackValueType::Array, true));  // message array
-      for (M const& val : vertexMessagePair.second) {
-        this->_format->addValue(data, val);
-        if (this->_sendToNextGSS) {
-          this->_sendCountNextGSS++;
-        } else {
-          this->_sendCount++;
+    VPackBuilder messagesVPack(&options);
+    {
+      VPackArrayBuilder ab(&messagesVPack);
+      for (auto const& [vertex, messages] : vertexMessageMap) {
+        messagesVPack.add(VPackValue(vertex));  // key
+        {
+          VPackArrayBuilder ab2(&messagesVPack);
+          for (M const& message : messages) {
+            this->_format->addValue(messagesVPack, message);
+            if (this->_sendToNextGSS) {
+              this->_sendCountNextGSS++;
+            } else {
+              this->_sendCount++;
+            }
+          }
```

Comment:
Can we use `serialize` for this?

Reply:
Yes, I would like to, but it is a bit more complicated, therefore I would like to do it in a separate PR: 
Messages are of generic type M. Each algorithm defines the type M and uses a Format class of type M to do serialization and deserialization. Therefore in the PregelMessage struct we have to make sure that we always use the correct message type.
Additionally we have different types of things we send around: either an array of messages to one vertex (using ArrayOutCache and ArrayInCache) or one message to one vertex (using CombiningOutCache and CombiningInCache). This difference has to be taken into account inside the PregelMessage if we don't use a basic VPackBuilder.
We should definitely do that, but I think it is a bit more work and should be done in a separate PR.